File: algorithms/js/36.有效的数独.js
/*
 * @lc app=leetcode.cn id=36 lang=javascript
 *
 * [36] 有效的数独
 */

// @lc code=start

const isValidPos = (row, col, board) => {
  let nums = [];
  const len = board.length;
  // check curent row
  for (let i = 0; i < len; i++) {
    const n = board[row][i];
    if (n !== ".") {
      if (nums.indexOf(n) !== -1) return false;
      nums.push(n);
    }
  }
  nums = [];
  // check curent col
  for (let i = 0; i < len; i++) {
    const n = board[i][col];
    if (n !== ".") {
      if (nums.indexOf(n) !== -1) return false;
      nums.push(n);
    }
  }
  nums = [];
  // check curent small-board
  let row_from = Math.floor(row / 3) * 3,
    col_from = Math.floor(col / 3) * 3;
  for (let i = row_from; i < row_from + 3; i++) {
    for (let j = col_from; j < col_from + 3; j++) {
      const n = board[i][j];
      if (n !== ".") {
        if (nums.indexOf(n) !== -1) return false;
        nums.push(n);
      }
    }
  }

  return true;
};
/**
 * @param {character[][]} board
 * @return {boolean}
 */

const isValidSudoku = (board) => {
  const len = board.length;
  for (let row = 0; row < len; row++) {
    for (let col = 0; col < len; col++) {
      if (!isValidPos(row, col, board)) return false;
    }
  }
  return true;
};
// @lc code=end

File: algorithms/golang/36.有效的数独.go
/*
 * @lc app=leetcode.cn id=36 lang=golang
 *
 * [36] 有效的数独
 */

// @lc code=start

/**
 * 分别记录每一行、每一列、每一区块出现的数字的个数。
 * 若出现次数大于1则为 invalid
 * 时间复杂度O(1)，空间复杂度O(1)
 */
func isValidSudoku(board [][]byte) bool {
	blocks := [3][3][9]int{}
	rows := [9][9]int{}
	cols := [9][9]int{}
	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			ch := board[i][j]
			if ch != '.' {
				idx := ch - '0' - 1
				blocks[i/3][j/3][idx]++
				rows[i][idx]++
				cols[j][idx]++
				if blocks[i/3][j/3][idx] > 1 || rows[i][idx] > 1 || cols[j][idx] > 1 {
					return false
				}
			}
		}
	}
	return true
}

// @lc code=end